ASM-7905 Puppet provider for server onboarding

kaushalgala opened the pull request:
Implementation for new type and provider for configuring credentials
and network settings for a given server using racadm commands. We keep
same style of implementation as we did for blades in dell-cmc module,
except for the fact the racadm calls here go directly to the server.
We also clean up and yard doc some execution methods in racadm.rb

Note: Maybe we should later refactor the code to move racadm command
to a common project where eventually all racadm calls for servers go to
single place. Right now some code is unnecessarily duplicated between
racadm.rb in dell-cmc and dell-idrac modules.

kaushalgala commented:
Still working on unit tests, as seem to be having difficulties with dependency module on transport for my tests.  In meanwhile, please review the code and provide feedbacks

gavin-scott commented on lib/puppet/provider/idrac_racadm.rb:
```diff
@@ -27,40 +27,68 @@ def client
+  def racadm_cmd(subcommand, params=[], process_opts={})
+    cmd = "racadm %s" % subcommand
+    unless params.empty?
+      param_string = params.is_a?(Array) ? params.join(" ") : params.to_s
```
`Array(params)` will do the "if array return the array otherwise make an array" check for you. So this block could be e.g.:

    cmd << " %s" % Array(params).join(" ") unless params.empty?

gavin-scott commented on lib/puppet/provider/idrac_racadm.rb:
```diff
@@ -27,40 +27,68 @@ def client
+    Puppet.info("racadm %s result: %s" % [subcommand, output])
+    if output.include?("ERROR:")
+      msg = "Error in racadm command %s" % subcommand
+      Puppet.err msg
```
Please include the parentheses around the arguments: `Puppet.err(msg)`

gavin-scott commented on lib/puppet/provider/idrac_racadm.rb:
```diff
@@ -27,40 +27,68 @@ def client
+      Puppet.err msg
+      raise msg if process_opts[:raise_on_err]
+    else
+      Puppet.info "Successfully executed %s" % subcommand
```
Parentheses

gavin-scott commented on lib/puppet/provider/server_onboard/default.rb:
```diff
@@ -0,0 +1,143 @@
+    rounds = (max_wait / sleep_interval)
+    connected = false
+
+    cmd = ["sudo", "wsman", "identify", "-h", ip, "-P", "443",
```
No need to use `sudo` here.

gavin-scott commented on lib/puppet/provider/server_onboard/default.rb:
```diff
@@ -0,0 +1,143 @@
+    connected = false
+
+    cmd = ["sudo", "wsman", "identify", "-h", ip, "-P", "443",
+           "-u", username, "-p", password, "-c", "dummy.cert", "-y", "basic", "-V", "-v"]
```
It would probably be a good idea to include `--transport-timeout` here as well. It might be worth adding this "identify" functionality into dell-asm-util's ASM::WsMan class and then we won't have to duplicate the arg setup here.

kaushalgala replied:
yeah, will move wait_for_ip to dell-asm-util

gavin-scott commented on lib/puppet/type/server_onboard.rb:
```diff
@@ -0,0 +1,29 @@
+    desc "certname of the server"
+  end
+
+  newproperty(:credential) do
```
It would be better to use `username` and `password` instead of `credential` since the credential is an internal ASM concept.

kaushalgala replied:
@gavin-scott to keep it consistent with other places ( e.g. dell-cmc idrac onboard, etc. ) I used credential.  We can discuss on Monday.

gavin-scott commented on lib/puppet/type/server_onboard.rb:
```diff
@@ -0,0 +1,29 @@
+  #
+  # specification for network configuration like static IP, DNS settings
+  #
+  # @param networks [Hash] Options for specifying the network config
```
An example of this data would be good. I guess that could be an `@example` tag here. Not sure how the YARD stuff will interact with puppet types.

kaushalgala replied:
ok will do

kaushalgala commented:
@gavin-scott  addressed all your comments, and also have uploaded my unit tests on the PR.

Like I mentioned in my email, the tests work with rspec, but having trouble with  rake execution. Will need your help on Monday

gavin-scott commented on .travis.yml:
```diff
@@ -1,5 +1,5 @@
 language: ruby
-script: "bundle exec rake lint && bundle exec rake"
+script: "bundle exec rake lint && bundle exec rake spec"
```
Should we just change the default task to "spec" instead of "rspec"? e.g. https://github.com/dell-asm/dell-idrac/blob/82a3f2226f9847dece254444bbe6322cbbf90aca/Rakefile#L34 and https://github.com/dell-asm/dell-idrac/blob/82a3f2226f9847dece254444bbe6322cbbf90aca/Rakefile#L40 Otherwise the tests will fail for a developer who just runs "bundle exec rake" correct?

kaushalgala replied:
let me try

gavin-scott commented on lib/puppet/type/server_onboard.rb:
```diff
@@ -0,0 +1,62 @@
+  #    credential = {
+  #      "username" => "foo",
+  #      "password" => "bar"
+  #    }
```
Nice, looks good.
